aten/src/ATen/native/sparse/cuda/SoftMax.cu

```diff
@@ -0,0 +1,680 @@
+
+#include <thrust/binary_search.h>
+#include <thrust/device_ptr.h>
+#include <thrust/sequence.h>
+#include <thrust/sort.h>
+#include <thrust/system/cuda/execution_policy.h>
+
+#include <cuda_runtime_api.h>
+#include <cusparse.h>
+#include <bitset>
+
+#include <c10/cuda/CUDAMathCompat.h>
+#include <ATen/cuda/CUDAApplyUtils.cuh>
+#include <ATen/cuda/detail/IndexUtils.cuh>
+#include <ATen/cuda/detail/OffsetCalculator.cuh>
+#include <ATen/native/cuda/Loops.cuh>
+
+#include <c10/macros/Macros.h>
+#include <thrust/copy.h>
+#include <thrust/device_ptr.h>
+#include <thrust/for_each.h>
+#include <thrust/functional.h>
+#include <thrust/gather.h>
+#include <thrust/generate.h>
+#include <thrust/iterator/discard_iterator.h>
+#include <thrust/reduce.h>
+#include <thrust/scan.h>
+#include <thrust/sequence.h>
+#include <thrust/sort.h>
+#include <thrust/system/cuda/execution_policy.h>
+#include <thrust/transform.h>
+#include <thrust/unique.h>
+
+#include <c10/cuda/CUDAMathCompat.h>
+
+namespace at {
+namespace native {
+namespace {
+
+// Number of threads in a block given an input size up to MAX_BLOCK_SIZE
+static int getNumThreads(int nElem) {
+#if defined(__HIP_PLATFORM_HCC__)
+  int threadSizes[5] = {16, 32, 64, 128, 256};
+#else
+  int threadSizes[5] = {32, 64, 128, 256, 512};
+#endif
+  for (int i = 0; i != 5; ++i) {
+    if (nElem <= threadSizes[i]) {
+      return threadSizes[i];
+    }
+  }
+  return threadSizes[4];
+}
+
+int64_t get_nvalues(const IntArrayRef& sizes, int64_t sparse_dim) {
+  /* Return the number of entries in the dense part of a sparse tensor.
+
+      `sizes` is a vector of sparse tensor dimensions.
+      `sparse_dim` is the dimension of the sparse part of a sparse tensor.
+    */
+  auto dim = sizes.size();
+  int64_t nvalues = 1;
+  for (auto i = sparse_dim; i < dim; i++) {
+    nvalues *= sizes[i];
+  }
+  return nvalues;
+}
+
+template <typename scalar_t, bool LogSoftMax>
+__global__ void cuda_sparse_coo_softmax_kernel(
+    int64_t* sorted_pool_indices,
+    int64_t size,
+    int64_t* pool_sizes,
+    int64_t* pool_offsets,
+    int64_t nvalues,
+    scalar_t* mx_rows,
+    scalar_t* exp_sums_rows,
+    PackedTensorAccessor<scalar_t, 2> input_values_acc,
+    PackedTensorAccessor<scalar_t, 2> ouput_values_acc) {
+  /*
+    See ATen/native/sparse/cpu/Softmax.cpp:cpu_sparse_coo_softmax for the CPU
+    implementation of the sparse softmax algorithm that this implementation is
+    based on.
+  */
+  int tid = threadIdx.x;
+  int blkid = blockIdx.x;
+  int blksz = blockDim.x;
+  int gridsz = gridDim.x;
+
+  int index = tid + blkid * blksz;
+  int step = blksz * gridsz;
+
+  while (index < size) {
+    int64_t offset = pool_offsets[index];
+    int64_t* pool_indices = sorted_pool_indices + offset;
+    int64_t pool_indices_size = pool_sizes[index];
+    scalar_t* mx_row = mx_rows + index * nvalues;
+
+    /* Apply exp to (v - mx) and sum the results */
+    scalar_t* exp_sums_row = exp_sums_rows + index * nvalues;
```

Comment:
Seems like exp sums can be accumulated in registers and then the final value written out; this might let you start with exp_sums_row = 0 instead of reading it (see previous comment about filling vs. leaving exp_sums_row uninitialized)